segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,413 @@
+                offset, sourceSegment);
+
+        Path sourcePath = Paths.get(config.getFilesystemRoot(), sourceSegment);
+        Path targetPath = Paths.get(config.getFilesystemRoot(), targetHandle.getSegmentName());
+
+        try (FileChannel targetChannel = (FileChannel) Files.newByteChannel(targetPath, EnumSet.of(StandardOpenOption.APPEND));
+             RandomAccessFile sourceFile = new RandomAccessFile(String.valueOf(sourcePath), "r")) {
+            if (isWritableFile(sourcePath)) {
+                throw new IllegalStateException( String.format("Source segment (%s) is not sealed.", sourceSegment));
+            }
+            long length = sourceFile.length();
+            while ( length > 0 ) {
+                long bytesTransferred = targetChannel.transferFrom(sourceFile.getChannel(),
+                        offset, length);
+                offset += bytesTransferred;
+                length -= bytesTransferred;
+            }
+            Files.delete(sourcePath);
+            LoggerHelpers.traceLeave(log, "concat", traceId);
+            return null;
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private Void syncDelete(SegmentHandle handle) {
+        Files.delete(Paths.get(config.getFilesystemRoot(), handle.getSegmentName()));
+        return null;
+    }
+
+    /**
+     * Executes the given supplier asynchronously and returns a Future that will be completed with the result.
+     */
+    private <R> CompletableFuture<R> supplyAsync(String segmentName, Supplier<R> operation) {
+        this.ensureInitializedAndNotClosed();
+
+        CompletableFuture<R> result = new CompletableFuture<>();
+        this.executor.execute(() -> {
+            try {
+                result.complete(operation.get());
+            } catch (Throwable e) {
+                handleException(e, segmentName, result);
+            }
+        });
+
+        return result;
+    }
+
+    private <R> void handleException(Throwable e, String segmentName, CompletableFuture<R> result) {
+        result.completeExceptionally(translateException(segmentName, e));
+    }
+
+    private Throwable translateException(String segmentName, Throwable e) {
+        Throwable retVal = e;
+        if (e instanceof NoSuchFileException || e instanceof FileNotFoundException) {
+            retVal = new StreamSegmentNotExistsException(segmentName);
+        }
+
+        if (e instanceof FileAlreadyExistsException) {
+            retVal = new StreamSegmentExistsException(segmentName);
+        }
+
+        if (e instanceof IndexOutOfBoundsException) {
+            retVal = new IllegalArgumentException(e.getMessage());
+        }
+
+        if (e instanceof AccessDeniedException) {
+            retVal = new StreamSegmentSealedException(segmentName, e);
+        }
+
+        return retVal;
+    }
+
+    private void ensureInitializedAndNotClosed() {
+        Exceptions.checkNotClosed(this.closed.get(), this);
```

Comment:
Since you're essentially only checking for closed, just get rid of this whole method and put `Exceptions.checkNotClosed` where you were invoking this.